packages/kolme-cli/src/chain_query.rs

```diff
@@ -0,0 +1,223 @@
+use anyhow::{anyhow, bail, ensure, Context, Result};
+use reqwest::{Client, Url};
+use serde::Deserialize;
+
+use crate::RequestBuilderExt;
+
+pub struct ChainApi {
+    client: Client,
+    url: Url,
+}
+
+#[derive(Deserialize)]
+#[allow(dead_code)]
+pub struct ChainVersion {
+    pub code_version: String,
+    pub chain_version: String,
+    pub next_height: BlockHeight,
+}
+
+#[allow(dead_code)]
+pub struct BlockResponse {
+    pub code_version: String,
+    pub chain_version: String,
+    pub block_height: BlockHeight,
+}
+
+pub struct ForkInfo {
+    pub first_block: BlockHeight,
+    pub last_block: BlockHeight,
+}
+
+#[derive(Deserialize, Copy, Clone, Debug)]
+pub struct BlockHeight(pub u32);
+
+impl BlockHeight {
+    pub fn pred(&self) -> Result<BlockHeight> {
+        let pred = self.0.checked_sub(1).context("Underflow in pred")?;
+        Ok(BlockHeight(pred))
+    }
+
+    pub fn succ(&self) -> Result<BlockHeight> {
+        let succ = self.0.checked_add(1).context("Overflow in succ")?;
+        Ok(BlockHeight(succ))
+    }
+
+    pub fn increasing_middle(&self, block_height: BlockHeight) -> Result<BlockHeight> {
+        ensure!(
+            self.0 <= block_height.0,
+            "Cannot compute middle since start_block is greater than end_block"
+        );
+
+        let middle = block_height
+            .0
+            .checked_sub(self.0)
+            .context("Overflow in sub")?
+            .checked_div(2)
+            .context("Underflow in div")?;
+
+        Ok(BlockHeight(middle + self.0))
+    }
+}
+
+impl ChainApi {
+    pub fn new(api_server: Url) -> Result<Self> {
+        let client = Client::builder().user_agent("kolme-cli").build()?;
+        Ok(ChainApi {
+            client,
+            url: api_server,
+        })
+    }
+
+    async fn root_info(&self) -> Result<ChainVersion> {
+        let response = self.client.get(self.url.clone()).send_check_json().await?;
+        Ok(response)
+    }
+
+    async fn block_response(&self, block: BlockHeight) -> Result<BlockResponse> {
+        #[derive(Deserialize)]
+        struct Response {
+            pub code_version: String,
+            pub chain_version: String,
+        }
+
+        let mut url = self.url.clone();
+        url.set_path(&format!("/block/{}", block.0));
+
+        let response: Response = self.client.get(url).send_check_json().await?;
+        let response = BlockResponse {
+            code_version: response.code_version,
+            chain_version: response.chain_version,
+            block_height: block,
+        };
+        Ok(response)
+    }
+
+    /// Find an arbitrary block height with a particular chain version
+    async fn find_block_height(&self, chain_version: &str) -> Result<BlockResponse> {
```

Comment:
Here's an idea... what about moving the logic for this entirely into the API server itself? It would require far less network traffic to pull off this search, likely speeding up the query significantly without any significant extra strain on the server (or, perhaps, even _less_ strain on the server).